RealmSwift/SwiftUI.swift

```diff
@@ -645,6 +645,285 @@ extension EnvironmentValues {
     }
 }
 
+@available(iOS 13.0, macOS 10.15, tvOS 13.0, watchOS 6.0, *)
+private class ObservableAsyncOpenStorage: ObservableObject {
+    var cancellables = [AnyCancellable]()
+    var realm: Realm?
+
+    @Published var asyncOpenState: AsyncOpenState = .notOpen {
+        willSet {
+            objectWillChange.send()
+        }
+    }
+
+    func asyncOpen(configuration: Realm.Configuration) -> RealmPublishers.AsyncOpenPublisher {
+        return Realm.asyncOpen(configuration: configuration)
+            .onProgressNotification { asyncProgress in
+                let progress = Progress(totalUnitCount: 1)
+                if asyncProgress.isTransferComplete {
+                    progress.completedUnitCount = Int64(1)
+                    self.asyncOpenState = .progress(progress)
+                } else {
+                    progress.completedUnitCount = Int64(asyncProgress.fractionTransferred)
+                    self.asyncOpenState = .progress(progress)
+                }
+            }
+    }
+}
+
+/**
+`AsyncOpenState`is an enum representing different states from `AsyncOpen` and `AutoOpen` process
+*/
+public enum AsyncOpenState {
+    /// Initial state
```

Comment:
This needs more elaboration. I assume it's the state we're in while we're waiting for a user to be logged in or while attempting to establish the connection? It might also be worth splitting those into `.waitingForUser` and `.connecting` if that's easy to do.